File: apps/backend/src/modules/addon/addon.dto.ts
import { IsArray, IsDate, IsObject, IsString } from 'class-validator';

import { AddonEnvDto, AddonEnvRequestDto } from '../addon-env/addon-env.dto';
import {
  AddonVersionDto,
  AddonVersionRequestDto,
} from '../addon-version/addon-version.dto';
import {
  AddonVolumeDto,
  AddonVolumeRequestDto,
} from '../addon-volume/addon-volume.dto';

import { Addon } from './addon.entity';

export class AddonDto {
  @IsString()
  id: string;

  @IsString()
  name: string;

  @IsString()
  description: string;

  @IsString()
  logo: string;

  @IsObject()
  versions: AddonVersionDto;

  @IsArray()
  volumes: AddonVolumeDto[];

  @IsArray()
  envs: AddonEnvDto[];

  @IsDate()
  creationDate: Date;

  @IsDate()
  updateTime: Date;

  constructor(addon: Addon) {
    this.id = addon.id;
    this.name = addon.name;
    this.description = addon.description;
    this.logo = addon.logo;
    this.versions = addon.versions;
    this.volumes = addon.volumes.getItems();
    this.envs = addon.envs.getItems();
    this.creationDate = addon.creationDate;
    this.updateTime = addon.updateTime;
  }
}

export class AddonRequestDto {
  @IsString()
  name: string;

  @IsString()
  description: string;

  @IsString()
  logo: string;

  @IsObject()
  versions: AddonVersionRequestDto;

  @IsArray()
  volumes: AddonVolumeRequestDto[];

  @IsArray()
  envs: AddonEnvRequestDto[];
}

File: backend/src/main/java/com/esmo/empaas/dtos/AddonDto.java
package com.esmo.empaas.dtos;

import java.util.List;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;

@Data
@AllArgsConstructor
@NoArgsConstructor
@ToString
public class AddonDto {

    private String id;
    private String name;
    private String fancyName;
    private String image;
    private String description;
    private String logo;
    private int port;
    private List<AddonVersionDto> versions;
    private List<AddonEnvDto> envs;
    private List<AddonVolumeDto> volumes;
}
